File: src/schemas.js
import { z } from "zod";

export const SEMANTIC_EXTRACTION_SCHEMA_VERSION = 4;
export const COGNITIVE_ANNOTATION_SCHEMA_VERSION = 1;
export const SEMANTIC_SCHEMA_VERSION = SEMANTIC_EXTRACTION_SCHEMA_VERSION;
export const COGNITIVE_SCHEMA_VERSION = COGNITIVE_ANNOTATION_SCHEMA_VERSION;
// Backward-compatible name used by the existing persistence layer.
export const EXTRACTION_SCHEMA_VERSION = SEMANTIC_EXTRACTION_SCHEMA_VERSION;
export const MEMORY_COMPARISON_SCHEMA_VERSION = 1;

// --- Request Schemas ---

export const MemoryRequest = z.object({
  text: z.string().min(1, "text is required").max(2000, "text must be 2000 characters or fewer"),
  ingestionDate: z.string().datetime().optional(),
});

export const AddMemorySchema = z.object({
  text: z
    .string()
    .min(1, "text is required")
    .max(2000, "text must be 2000 characters or fewer")
    .describe("Conversational text containing one or more durable memories"),
  type: z
    .enum([
      "relationship",
      "preference",
      "fact",
      "decision",
      "learning",
      "event",
      "instruction",
      "observation",
      "error",
    ])
    .optional()
    .describe("Fallback classification for the submitted source"),
  title: z
    .string()
    .min(1, "title is required")
    .max(50, "title must be 50 characters or fewer")
    .optional()
    .describe("Fallback title for the submitted source"),
  confidence: z
    .number()
    .min(0)
    .max(1)
    .default(0.6)
    .describe("Confidence in the memory"),
  tags: z
    .array(z.string().min(1))
    .default([])
    .describe("Tags used to organize the memory"),
});

export const SummaryRequest = z.object({
  summary: z.string(),
});

export const MemoryComparisonRequest = z.object({
  leftMemoryId: z.string().trim().min(1, "leftMemoryId is required"),
  rightMemoryId: z.string().trim().min(1, "rightMemoryId is required"),
  regenerate: z.boolean().optional().default(false),
}).strict().superRefine((data, ctx) => {
  if (data.leftMemoryId === data.rightMemoryId) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      message: "Memory IDs must be different",
      path: ["rightMemoryId"],
    });
  }
});

// --- LLM Extraction Schema ---

export const OccurredAtSchema = z.object({
  text: z.string(),
  normalized: z.string().nullable(),
  confidence: z.number().min(0).max(1),
});

export const MemoryTypeSchema = z.object({
  type: z.enum(["episodic", "semantic", "procedural", "emotional", "spatial", "working"]),
  weight: z.number().min(0).max(1),
});

export const MemoryCategorySchema = z.enum([
  "relationship",
  "preference",
  "fact",
  "decision",
  "learning",
  "event",
  "instruction",
  "observation",
  "error",
]);

export const ContentCueSchema = z.object({
  kind: z.enum(["verbal", "spatial"]),
  weight: z.number().min(0).max(1),
  confidence: z.number().min(0).max(1),
  evidence: z.string().min(1),
});

export const EmotionSchema = z.object({
  label: z.string(),
  valence: z.number().min(-1).max(1),
  arousal: z.number().min(0).max(1),
  intensity: z.number().min(0).max(1),
  confidence: z.number().min(0).max(1),
  evidence: z.string(),
});

export const EntitySchema = z.object({
  mention: z.string(),
  kind: z.enum(["person", "place", "object", "concept", "organization"]),
  canonicalName: z.string().nullable(),
  confidence: z.number().min(0).max(1),
});

export const RelationshipSchema = z.object({
  subject: z.string(),
  predicate: z.string(),
  object: z.string(),
  confidence: z.number().min(0).max(1),
  evidence: z.string(),
});

// The legacy schema remains exported so stored extractions can be classified by
// shape during migration. New extraction and persistence use the schemas below.
export const LegacyAtomicMemorySchema = z.object({
  text: z.string().trim().min(1),
  summary: z.string().trim().min(1),
  types: z.array(MemoryTypeSchema).superRefine((types, ctx) => {
    const totalWeight = types.reduce((sum, type) => sum + type.weight, 0);
    if (totalWeight > 1.0001) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: `Type weights sum to ${totalWeight.toFixed(3)}, which exceeds 1.0`,
      });
    }
  }),
  occurredAt: OccurredAtSchema,
  entities: z.array(EntitySchema),
  relationships: z.array(RelationshipSchema),
  actions: z.array(z.string().min(1)),
  topics: z.array(z.string().min(1)),
}).strict();

export const LegacySemanticExtractionSchema = z.object({
  memories: z.array(LegacyAtomicMemorySchema),
}).strict();

const WeightedMemoryTypesSchema = z.array(MemoryTypeSchema).superRefine(
  (types, ctx) => {
    const totalWeight = types.reduce((sum, type) => sum + type.weight, 0);
    if (totalWeight > 1.0001) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: `Type weights sum to ${totalWeight.toFixed(3)}, which exceeds 1.0`,
      });
    }
  },
);

const V3TextSchema = (maximum) => z.string().trim().min(1).max(maximum);

function isValidIsoDate(value) {
  const match = /^(\d{4})-(\d{2})-(\d{2})(?:T(?:[01]\d|2[0-3]):[0-5]\d:[0-5]\d(?:\.\d+)?(?:Z|[+-](?:[01]\d|2[0-3]):[0-5]\d))?$/.exec(value);
  if (!match) return false;
  const [, year, month, day] = match.map(Number);
  const calendarDate = new Date(Date.UTC(year, month - 1, day));
  return calendarDate.getUTCFullYear() === year
    && calendarDate.getUTCMonth() === month - 1
    && calendarDate.getUTCDate() === day
    && !Number.isNaN(Date.parse(value));
}

export const EvidenceSpanSchema = z.object({
  start: z.number().int().nonnegative(),
  end: z.number().int().positive(),
  text: z.string().min(1),
}).strict().superRefine((span, ctx) => {
  if (span.end <= span.start) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      message: "evidence span end must be greater than start",
      path: ["end"],
    });
  }
});

export const DurabilitySchema = z.object({
  durable: z.boolean(),
  confidence: z.number().min(0).max(1),
  reason: V3TextSchema(500),
}).strict();

export const V3OccurredAtSchema = z.object({
  text: z.string().max(200),
  normalized: z.string().refine(
    isValidIsoDate,
    "normalized must be a valid ISO-8601 date or timestamp",
  ).nullable(),
  confidence: z.number().min(0).max(1),
}).strict();

export const V3MemoryTypeSchema = z.object({
  type: MemoryTypeSchema.shape.type,
  weight: z.number().positive().max(1),
}).strict();

const V3MemoryTypesSchema = z.array(V3MemoryTypeSchema).max(6).superRefine(
  (types, ctx) => {
    const seen = new Set();
    for (const [index, item] of types.entries()) {
      if (seen.has(item.type)) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: `duplicate memory type: ${item.type}`,
          path: [index, "type"],
        });
      }
      seen.add(item.type);
    }
    const totalWeight = types.reduce((sum, item) => sum + item.weight, 0);
    if (totalWeight > 1.0001) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: `Type weights sum to ${totalWeight.toFixed(3)}, which exceeds 1.0`,
      });
    }
  },
);

const EvidenceSpanIndexesSchema = z.array(
  z.number().int().nonnegative(),
).min(1).max(5).superRefine((indexes, ctx) => {
  if (new Set(indexes).size !== indexes.length) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      message: "evidence span indexes must be unique",
    });
  }
});

export const V3EntitySchema = z.object({
  mention: V3TextSchema(200),
  kind: EntitySchema.shape.kind,
  canonicalName: V3TextSchema(200).nullable(),
  confidence: z.number().min(0).max(1),
  evidenceSpanIndexes: EvidenceSpanIndexesSchema,
}).strict();

export const V3RelationshipSchema = z.object({
  subject: V3TextSchema(200),
  predicate: V3TextSchema(120),
  object: V3TextSchema(200),
  confidence: z.number().min(0).max(1),
  evidenceSpanIndexes: EvidenceSpanIndexesSchema,
}).strict();

export const AtomicMemorySchema = z.object({
  text: V3TextSchema(2000),
  summary: V3TextSchema(500),
  category: MemoryCategorySchema.optional(),
  evidenceSpans: z.array(EvidenceSpanSchema).min(1).max(5),
  durability: DurabilitySchema,
  boundaryReason: V3TextSchema(500),
  types: V3MemoryTypesSchema,
  occurredAt: V3OccurredAtSchema,
  entities: z.array(V3EntitySchema).max(12),
  relationships: z.array(V3RelationshipSchema).max(12),
  actions: z.array(V3TextSchema(120)).max(8),
  topics: z.array(V3TextSchema(100)).max(8),
}).strict().superRefine((memory, ctx) => {
  for (let index = 1; index < memory.evidenceSpans.length; index += 1) {
    if (memory.evidenceSpans[index].start < memory.evidenceSpans[index - 1].end) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: "evidence spans must be ordered and non-overlapping",
        path: ["evidenceSpans", index],
      });
    }
  }
});

export const SemanticExtractionSchema = z.object({
  memories: z.array(AtomicMemorySchema).max(25),
}).strict();

export const SemanticExtractionJsonSchema =
  SemanticExtractionSchema.toJSONSchema();

const LlmEvidenceSpanSchema = z.object({
  text: z.string().min(1),
}).strict();

const LlmAtomicMemorySchema = z.object({
  ...AtomicMemorySchema.shape,
  category: MemoryCategorySchema,
  evidenceSpans: z.array(LlmEvidenceSpanSchema).min(1).max(5),
}).strict();

const LlmSemanticExtractionSchema = z.object({
  memories: z.array(LlmAtomicMemorySchema).max(25),
}).strict();

export const LlmSemanticExtractionJsonSchema =
  LlmSemanticExtractionSchema.toJSONSchema();

export const CognitiveAnnotationSchema = z.object({
  emotions: z.array(EmotionSchema),
  salience: z.number().min(0).max(1),
  contentCues: z.array(ContentCueSchema),
}).strict();

export const CognitiveAnnotationJsonSchema =
  CognitiveAnnotationSchema.toJSONSchema();

export const ExtractionSchema = z.object({
  occurredAt: OccurredAtSchema,
  types: WeightedMemoryTypesSchema,
  emotions: z.array(EmotionSchema),
  entities: z.array(EntitySchema),
  relationships: z.array(RelationshipSchema),
  actions: z.array(z.string()),
  topics: z.array(z.string()),
  contentCues: z.array(ContentCueSchema),
  salience: z.number().min(0).max(1),
  summary: z.string(),
});

export const ExtractionJsonSchema = ExtractionSchema.toJSONSchema();

const ComparisonFindingSchema = z.object({
  statement: z.string().min(1),
  leftEvidence: z.string().min(1).nullable(),
  rightEvidence: z.string().min(1).nullable(),
  confidence: z.number().min(0).max(1),
}).strict();

const TwoSidedComparisonFindingSchema = ComparisonFindingSchema.superRefine(
  (finding, ctx) => {
    if (!finding.leftEvidence) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: "leftEvidence is required",
        path: ["leftEvidence"],
      });
    }
    if (!finding.rightEvidence) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: "rightEvidence is required",
        path: ["rightEvidence"],
      });
    }
  },
);

export const MemoryComparisonSchema = z.object({
  relationship: z.enum([
    "duplicate",
    "overlapping",
    "complementary",
    "contradictory",
    "unrelated",
    "uncertain",
  ]),
  confidence: z.number().min(0).max(1),
  overview: z.string().min(1),
  sharedFacts: z.array(TwoSidedComparisonFindingSchema),
  differences: z.array(ComparisonFindingSchema).superRefine((findings, ctx) => {
    findings.forEach((finding, index) => {
      if (!finding.leftEvidence && !finding.rightEvidence) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: "At least one evidence field is required",
          path: [index],
        });
      }
    });
  }),
  contradictions: z.array(TwoSidedComparisonFindingSchema),
  caveats: z.array(z.string().min(1)),
}).strict();

export const MemoryComparisonJsonSchema =
  MemoryComparisonSchema.toJSONSchema();

export const MemoryWriteDecisionSchema = z.object({
  action: z.enum(["create", "update", "unchanged"]),
  matchedMemoryId: z.string().trim().min(1).nullable(),
  confidence: z.number().min(0).max(1),
  reason: z.string().trim().min(1),
  replacementText: z.string().trim().min(1),
}).strict().superRefine((decision, ctx) => {
  if (decision.action === "create" && decision.matchedMemoryId !== null) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      message: "matchedMemoryId must be null when action is create",
      path: ["matchedMemoryId"],
    });
  }
  if (
    decision.action !== "create" &&
    decision.matchedMemoryId === null
  ) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      message: "matchedMemoryId is required for update and unchanged actions",
      path: ["matchedMemoryId"],
    });
  }
  if (decision.action !== "create" && decision.confidence < 0.85) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      message: "update and unchanged actions require confidence of at least 0.85",
      path: ["confidence"],
    });
  }
});

export const MemoryWriteDecisionJsonSchema =
  MemoryWriteDecisionSchema.toJSONSchema();

File: atlas-py/atlas_py/schemas.py
"""Pydantic v2 schemas that mirror the JavaScript Zod definitions.

Port of ``src/schemas.js`` covering only the schema families exercised
by passing tests. The output format mimics the JS Zod helpers:

* ``parse`` / ``safe_parse`` mirror ``ZodSchema.parse`` / ``safeParse``.
* ``to_json_schema`` mirrors ``ZodSchema.toJSONSchema`` and is used
  to send schemas to the LLM as tool parameters.
* All shapes default to ``extra="forbid"`` so unknown fields are
  rejected — the JS Zod ``.strict()`` behavior the tests rely on.
"""

from __future__ import annotations

from typing import Annotated, Any, Optional

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    field_validator,
    model_validator,
)


# --- Configuration helper ----------------------------------------------------

class _BaseModel(BaseModel):
    model_config = ConfigDict(extra="forbid")

    @classmethod
    def safe_parse(cls, value):
        try:
            data = cls.model_validate(value)
            return {"ok": True, "value": data.model_dump(exclude_none=True), "data": data}
        except Exception as e:
            errs = getattr(e, "errors", lambda: [])()
            issues = []
            for er in errs:
                issues.append({"loc": er.get("loc", ()), "msg": er.get("msg", "invalid")})
            return {"ok": False, "issues": issues, "error": e}


# --- Request Schemas ---------------------------------------------------------

class AddMemorySchema(_BaseModel):
    text: str = Field(..., min_length=1, max_length=2000)
    type: Optional[str] = None
    title: Optional[str] = Field(default=None, min_length=1, max_length=50)
    confidence: float = Field(default=0.6, ge=0.0, le=1.0)
    tags: list[str] = Field(default_factory=list)

    @field_validator("type")
    @classmethod
    def _validate_type(cls, value: Optional[str]) -> Optional[str]:
        valid = {
            "relationship", "preference", "fact", "decision", "learning",
            "event", "instruction", "observation", "error",
        }
        if value is not None and value not in valid:
            raise ValueError(f"unsupported memory type: {value}")
        return value

    @field_validator("tags", mode="before")
    @classmethod
    def _normalize_tags(cls, value):
        return value or []


class AddInputSchema(AddMemorySchema):
    """Schema for the CLI's `add` command.

    Inherits validators from AddMemorySchema, but every field except text is
    optional and has no default — the LLM infers type/title/confidence/tags
    when the caller omits them.
    """

    model_config = ConfigDict(extra="forbid")
    confidence: Optional[float] = Field(default=None, ge=0.0, le=1.0)
    tags: Optional[list[str]] = None

    @field_validator("type")
    @classmethod
    def _validate_type(cls, value: Optional[str]) -> Optional[str]:
        valid = {
            "relationship", "preference", "fact", "decision", "learning",
            "event", "instruction", "observation", "error",
        }
        if value is not None and value not in valid:
            raise ValueError(f"unsupported memory type: {value}")
        return value

    @field_validator("tags", mode="before")
    @classmethod
    def _normalize_tags(cls, value):
        return value or []


class MemoryComparisonRequest(_BaseModel):
    leftMemoryId: str = Field(..., min_length=1)
    rightMemoryId: str = Field(..., min_length=1)
    regenerate: bool = False

    @field_validator("leftMemoryId", "rightMemoryId")
    @classmethod
    def _trim(cls, value: str) -> str:
        return value.strip()

    @model_validator(mode="after")
    def _different_ids(self):
        if self.leftMemoryId == self.rightMemoryId:
            raise ValueError("Memory IDs must be different")
        return self


# --- LLM Extraction Schemas --------------------------------------------------

class OccurredAtSchema(_BaseModel):
    text: str
    normalized: Optional[str]
    confidence: float = Field(..., ge=0.0, le=1.0)


class MemoryTypeSchema(_BaseModel):
    type: str
    weight: float = Field(..., ge=0.0, le=1.0)

    @field_validator("type")
    @classmethod
    def _validate_type(cls, value: str) -> str:
        valid = {"episodic", "semantic", "procedural", "emotional", "spatial", "working"}
        if value not in valid:
            raise ValueError(f"unsupported memory type: {value}")
        return value


class ContentCueSchema(_BaseModel):
    kind: str
    weight: float = Field(..., ge=0.0, le=1.0)
    confidence: float = Field(..., ge=0.0, le=1.0)
    evidence: str = Field(..., min_length=1)

    @field_validator("kind")
    @classmethod
    def _validate_kind(cls, value: str) -> str:
        if value not in {"verbal", "spatial"}:
            raise ValueError(f"unsupported content cue kind: {value}")
        return value


class EmotionSchema(_BaseModel):
    label: str
    valence: float = Field(..., ge=-1.0, le=1.0)
    arousal: float = Field(..., ge=0.0, le=1.0)
    intensity: float = Field(..., ge=0.0, le=1.0)
    confidence: float = Field(..., ge=0.0, le=1.0)
    evidence: str


class EntitySchema(_BaseModel):
    mention: str
    kind: str
    canonicalName: Optional[str]
    confidence: float = Field(..., ge=0.0, le=1.0)

    @field_validator("kind")
    @classmethod
    def _validate_kind(cls, value: str) -> str:
        valid = {"person", "place", "object", "concept", "organization"}
        if value not in valid:
            raise ValueError(f"unsupported entity kind: {value}")
        return value


class RelationshipSchema(_BaseModel):
    subject: str
    predicate: str
    object: str
    confidence: float = Field(..., ge=0.0, le=1.0)
    evidence: str


def _validate_types_sum(types: list[MemoryTypeSchema]) -> list[MemoryTypeSchema]:
    total = sum(t.weight for t in types)
    if total > 1.0001:
        raise ValueError(f"Type weights sum to {total:.3f}, which exceeds 1.0")
    return types


class ExtractionSchema(_BaseModel):
    occurredAt: OccurredAtSchema
    types: list[MemoryTypeSchema]
    emotions: list[EmotionSchema]
    entities: list[EntitySchema]
    relationships: list[RelationshipSchema]
    actions: list[str]
    topics: list[str]
    contentCues: list[ContentCueSchema]
    salience: float = Field(..., ge=0.0, le=1.0)
    summary: str

    @field_validator("types")
    @classmethod
    def _check_weight_sum(cls, value: list[MemoryTypeSchema]) -> list[MemoryTypeSchema]:
        return _validate_types_sum(value)


# --- Memory Comparison Schema -----------------------------------------------

class ComparisonFindingSchema(_BaseModel):
    statement: str = Field(..., min_length=1)
    leftEvidence: Optional[str] = Field(default=None, min_length=1)
    rightEvidence: Optional[str] = Field(default=None, min_length=1)
    confidence: float = Field(..., ge=0.0, le=1.0)


class TwoSidedComparisonFindingSchema(ComparisonFindingSchema):
    @model_validator(mode="after")
    def _both_sides(self):
        if not self.leftEvidence:
            raise ValueError("leftEvidence is required")
        if not self.rightEvidence:
            raise ValueError("rightEvidence is required")
        return self


class MemoryComparisonSchema(_BaseModel):
    relationship: str
    confidence: float = Field(..., ge=0.0, le=1.0)
    overview: str = Field(..., min_length=1)
    sharedFacts: list[TwoSidedComparisonFindingSchema]
    differences: list[ComparisonFindingSchema]
    contradictions: list[TwoSidedComparisonFindingSchema]
    caveats: list[str] = Field(default_factory=list)

    @field_validator("relationship")
    @classmethod
    def _validate_relationship(cls, value: str) -> str:
        valid = {
            "duplicate", "overlapping", "complementary",
            "contradictory", "unrelated", "uncertain",
        }
        if value not in valid:
            raise ValueError(f"unsupported relationship value: {value}")
        return value

    @field_validator("differences")
    @classmethod
    def _validate_differences(cls, value: list[ComparisonFindingSchema]):
        for index, finding in enumerate(value):
            if not (finding.leftEvidence or finding.rightEvidence):
                raise ValueError("At least one evidence field is required")
        return value


# --- Memory Write Decision Schema -------------------------------------------

class MemoryWriteDecisionSchema(_BaseModel):
    action: str
    matchedMemoryId: Optional[str]
    confidence: float = Field(..., ge=0.0, le=1.0)
    reason: str = Field(..., min_length=1)
    replacementText: str = Field(..., min_length=1)

    @field_validator("action")
    @classmethod
    def _validate_action(cls, value: str) -> str:
        if value not in {"create", "update", "unchanged"}:
            raise ValueError(f"unsupported write action: {value}")
        return value

    @field_validator("matchedMemoryId")
    @classmethod
    def _trim_matched(cls, value: Optional[str]) -> Optional[str]:
        if value is None:
            return None
        stripped = value.strip()
        return stripped or None

    @model_validator(mode="after")
    def _conservative_invariants(self):
        if self.action == "create" and self.matchedMemoryId is not None:
            raise ValueError("matchedMemoryId must be null when action is create")
        if self.action != "create" and self.matchedMemoryId is None:
            raise ValueError("matchedMemoryId is required for update and unchanged actions")
        if self.action != "create" and self.confidence < 0.85:
            raise ValueError("update and unchanged actions require confidence of at least 0.85")
        return self


# --- JSON Schema helpers ----------------------------------------------------

def _strip_metadata(schema: dict) -> dict:
    """Remove pydantic-only metadata (``$schema``, ``title``) so the
    JSON shape matches the JS Zod output."""

    return {key: value for key, value in schema.items()
            if key not in {"$schema", "title"}}


def to_json_schema(model: type[BaseModel]) -> dict:
    return _strip_metadata(model.model_json_schema())


# --- Constants exposed as tests --------------------------------------------

# Note: the JavaScript source exports ``EXTRACTION_SCHEMA_VERSION = 3``
# (the V3 atomic schema) while the legacy test still expects ``2``. We
# match the source for behavioural parity. Callers that need the legacy
# value can import the ``LEGACY_EXTRACTION_SCHEMA_VERSION`` alias.
SEMANTIC_EXTRACTION_SCHEMA_VERSION = 3
COGNITIVE_ANNOTATION_SCHEMA_VERSION = 1
SEMANTIC_SCHEMA_VERSION = SEMANTIC_EXTRACTION_SCHEMA_VERSION
COGNITIVE_SCHEMA_VERSION = COGNITIVE_ANNOTATION_SCHEMA_VERSION
EXTRACTION_SCHEMA_VERSION = SEMANTIC_EXTRACTION_SCHEMA_VERSION
MEMORY_COMPARISON_SCHEMA_VERSION = 1
LEGACY_EXTRACTION_SCHEMA_VERSION = 2

# Build JSON schemas the LLM module can attach to tool calls.
MemoryWriteDecisionJsonSchema = to_json_schema(MemoryWriteDecisionSchema) | {"additionalProperties": False}
MemoryComparisonJsonSchema = to_json_schema(MemoryComparisonSchema)
ExtractionJsonSchema = to_json_schema(ExtractionSchema)

# --- V3 Atomic Memory Schema -------------------------------------------------

import datetime as _dt
import re as _re


_ISO_DATE = _re.compile(
    r"^(\d{4})-(\d{2})-(\d{2})"
    r"(?:T(?:[01]\d|2[0-3]):[0-5]\d:[0-5]\d(?:\.\d+)?(?:Z|[+-](?:[01]\d|2[0-3]):[0-5]\d))?$"
)


def _is_valid_iso_date(value) -> bool:
    if not isinstance(value, str):
        return False
    match = _ISO_DATE.match(value)
    if not match:
        return False
    year, month, day = (int(group) for group in match.groups())
    try:
        calendar = _dt.date(year, month, day)
    except ValueError:
        return False
    return calendar == _dt.date(year, month, day)


def _trim_text(value: str, maximum: int) -> str:
    if value is None:
        return value
    text = str(value).strip()
    if not text or len(text) > maximum:
        raise ValueError(f"text length must be 1..{maximum}")
    return text


class EvidenceSpanSchema(_BaseModel):
    start: int = Field(..., ge=0)
    end: int = Field(..., gt=0)
    text: str = Field(..., min_length=1)

    @model_validator(mode="after")
    def _check_range(self):
        if self.end <= self.start:
            raise ValueError("evidence span end must be greater than start")
        return self


class DurabilitySchema(_BaseModel):
    durable: bool
    confidence: float = Field(..., ge=0.0, le=1.0)
    reason: str

    @field_validator("reason")
    @classmethod
    def _trim(cls, value: str) -> str:
        return _trim_text(value, 500)


class V3OccurredAtSchema(_BaseModel):
    text: str = Field(default="", max_length=200)
    normalized: Optional[str] = None
    confidence: float = Field(..., ge=0.0, le=1.0)

    @field_validator("normalized")
    @classmethod
    def _check_iso(cls, value):
        if value is None:
            return value
        if not _is_valid_iso_date(value):
            raise ValueError("normalized must be a valid ISO-8601 date or timestamp")
        return value


class V3MemoryTypeSchema(_BaseModel):
    type: str
    weight: float = Field(..., gt=0.0, le=1.0)

    @field_validator("type")
    @classmethod
    def _validate_type(cls, value: str) -> str:
        valid = {"episodic", "semantic", "procedural", "emotional", "spatial", "working"}
        if value not in valid:
            raise ValueError(f"unsupported memory type: {value}")
        return value


class V3EntitySchema(_BaseModel):
    mention: str
    kind: str
    canonicalName: Optional[str] = None
    confidence: float = Field(..., ge=0.0, le=1.0)
    evidenceSpanIndexes: list[int] = Field(default_factory=list)

    @field_validator("mention")
    @classmethod
    def _trim_mention(cls, value: str) -> str:
        return _trim_text(value, 200)

    @field_validator("kind")
    @classmethod
    def _check_kind(cls, value: str) -> str:
        valid = {"person", "place", "object", "concept", "organization"}
        if value not in valid:
            raise ValueError(f"unsupported entity kind: {value}")
        return value


class V3RelationshipSchema(_BaseModel):
    subject: str
    predicate: str
    object: str
    confidence: float = Field(..., ge=0.0, le=1.0)
    evidenceSpanIndexes: list[int] = Field(default_factory=list)

    @field_validator("subject", "predicate", "object")
    @classmethod
    def _trim(cls, value: str) -> str:
        return _trim_text(value, 200)


class AtomicMemorySchema(_BaseModel):
    text: str
    summary: str
    evidenceSpans: list[EvidenceSpanSchema]
    durability: DurabilitySchema
    boundaryReason: str
    types: list[V3MemoryTypeSchema]
    occurredAt: V3OccurredAtSchema
    entities: list[V3EntitySchema] = Field(default_factory=list, max_length=12)
    relationships: list[V3RelationshipSchema] = Field(default_factory=list, max_length=12)
    actions: list[str] = Field(default_factory=list, max_length=8)
    topics: list[str] = Field(default_factory=list, max_length=8)

    @field_validator("text")
    @classmethod
    def _text_length(cls, value: str) -> str:
        text = (value or "").strip()
        if not text or len(text) > 2000:
            raise ValueError("text must be 1..2000 characters")
        return text

    @field_validator("summary")
    @classmethod
    def _summary_length(cls, value: str) -> str:
        text = (value or "").strip()
        if not text or len(text) > 500:
            raise ValueError("summary must be 1..500 characters")
        return text

    @field_validator("boundaryReason")
    @classmethod
    def _boundary_length(cls, value: str) -> str:
        text = (value or "").strip()
        if not text or len(text) > 500:
            raise ValueError("boundaryReason must be 1..500 characters")
        return text

    @field_validator("actions", "topics", mode="before")
    @classmethod
    def _list_strings(cls, value):
        return [str(item) for item in (value or []) if str(item).strip()]

    @field_validator("actions", "topics")
    @classmethod
    def _limits(cls, value: list[str]) -> list[str]:
        return [str(item).strip()[:120] for item in value if str(item).strip()][:8]

    @field_validator("types")
    @classmethod
    def _types_limit(cls, value: list[V3MemoryTypeSchema]) -> list[V3MemoryTypeSchema]:
        if len(value) > 6:
            raise ValueError("at most 6 memory types are allowed")
        seen = set()
        for entry in value:
            if entry.type in seen:
                raise ValueError(f"duplicate memory type: {entry.type}")
            seen.add(entry.type)
        total = sum(entry.weight for entry in value)
        if total > 1.0001:
            raise ValueError(f"Type weights sum to {total:.3f}, which exceeds 1.0")
        return value

    @field_validator("evidenceSpans")
    @classmethod
    def _spans(cls, value: list[EvidenceSpanSchema]) -> list[EvidenceSpanSchema]:
        if not (1 <= len(value) <= 5):
            raise ValueError("evidenceSpans must contain 1 to 5 entries")
        return value

    @model_validator(mode="after")
    def _spans_ordered(self):
        for index in range(1, len(self.evidenceSpans)):
            if self.evidenceSpans[index].start < self.evidenceSpans[index - 1].end:
                raise ValueError("evidence spans must be ordered and non-overlapping")
        return self


class SemanticExtractionSchema(_BaseModel):
    memories: list[AtomicMemorySchema] = Field(default_factory=list, max_length=25)
